Add EpiGEN new recipe

msantorsola opened the pull request:
This PR adds a recipe for [EpiGEN](https://github.com/biomedbigdata/epigen/tree/master), a tool for genotype and phenotype data simulation.

kbseah commented:
@msantorsola: I think it'll be better if you submit separate pull requests for `gametes` and `epigen`. It looks like they are two independent packages, right?

kbseah commented on recipes/epigen/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+  version: {{ version }}
+
+source:
+  url: https://github.com/biomedbigdata/epigen/archive/refs/heads/master.zip
```
the URL should point to a specific tag or release, ([ref](https://bioconda.github.io/contributor/guidelines.html#stable-url)). otherwise this recipe will break if there are ever any new commits to the master branch. it looks like this repo hasn't been updated in a while, but better to be on the safe side. you can also point to a specific commit [by its hash](https://docs.github.com/en/repositories/working-with-files/using-files/downloading-source-code-archives#source-code-archive-urls)

msantorsola replied:
thank you @kbseah. I will update the recipe to point to a specific commit as you recommended.

msantorsola commented:
> @msantorsola: I think it'll be better if you submit separate pull requests for `gametes` and `epigen`. It looks like they are two independent packages, right?

To solve this issue, I think that the best solution is to create a new branch for epigen, and then a new PR. If there's a preference for a different approach or additional steps that should be considered,  I'll be happy to accommodate.

kbseah commented:
> To solve this issue, I think that the best solution is to create a new branch for epigen, and then a new PR. If there's a preference for a different approach or additional steps that should be considered, I'll be happy to accommodate.

@msantorsola : i think you mean to make a new branch and PR for `gametes`, right? since this PR has "epigen" in the title, it should have only the commits for the epigen recipe. i'll be happy to review the other PR too so please feel free to tag me in the new PR when you open it.

msantorsola commented:
> @msantorsola: I think it'll be better if you submit separate pull requests for `gametes` and `epigen`. It looks like they are two independent packages, right?


> > To solve this issue, I think that the best solution is to create a new branch for epigen, and then a new PR. If there's a preference for a different approach or additional steps that should be considered, I'll be happy to accommodate.
> 
> @msantorsola : i think you mean to make a new branch and PR for `gametes`, right? since this PR has "epigen" in the title, it should have only the commits for the epigen recipe. i'll be happy to review the other PR too so please feel free to tag me in the new PR when you open it.

Dear @kbseah, there is an existing PR for `gametes` you can find [here](https://github.com/bioconda/bioconda-recipes/pull/43783), which I have locally tested, and seems to work. 
The issue arises with the `epigen` branch which, as you rightly pointed out, seems to have been branched from `gametes` instead of the master, also including changes specific to `gametes`.
So, I could create a new branch for `epigen`, to ensure no more overlap with `gametes`, and open a new PR for the epigen changes, while retaining the active gametes PR. What do you think about?

kbseah commented:
@msantorsola thanks for the clarification, i didn't see that there was already a PR open for gametes.

alternatively you could just delete the gametes recipe files from this branch, since the files are already in the other branch.

kbseah commented:
I'm unsure about the license here, and have posted in the BioConda Gitter channel asking for input

msantorsola commented:
> I'm unsure about the license here, and have posted in the BioConda Gitter channel asking for input

OK, thank you @kbseah . I'm testing it locally now, but I still have issues importing the tool

```
>>> import sys
>>> print(sys.path)
['', '/home/santorsola/.conda/envs/test_epigen/lib/python310.zip', '/home/santorsola/.conda/envs/test_epigen/lib/python3.10', '/home/santorsola/.conda/envs/test_epigen/lib/python3.10/lib-dynload', '/home/santorsola/.local/lib/python3.10/site-packages', '/home/santorsola/.conda/envs/test_epigen/lib/python3.10/site-packages']


```

```
$conda list
# packages in environment at /home/santorsola/.conda/envs/test_epigen:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       2_gnu    conda-forge
bzip2                     1.0.8                hd590300_5    conda-forge
ca-certificates           2023.11.17           hbcca054_0    conda-forge
epigen                    1               py312hdfd78af_0    file:///home/santorsola/.conda/envs/bioconda/conda-bld
ld_impl_linux-64          2.40                 h41732ed_0    conda-forge
libffi                    3.4.2                h7f98852_5    conda-forge
libgcc-ng                 13.2.0               h807b86a_3    conda-forge
libgomp                   13.2.0               h807b86a_3    conda-forge
libnsl                    2.0.1                hd590300_0    conda-forge
libsqlite                 3.44.2               h2797004_0    conda-forge
libuuid                   2.38.1               h0b41bf4_0    conda-forge
libxcrypt                 4.4.36               hd590300_1    conda-forge
libzlib                   1.2.13               hd590300_5    conda-forge
ncurses                   6.4                  h59595ed_2    conda-forge
openssl                   3.2.0                hd590300_1    conda-forge
pip                       23.3.2             pyhd8ed1ab_0    conda-forge
python                    3.10.13         hd12c33a_1_cpython    conda-forge
readline                  8.2                  h8228510_1    conda-forge
setuptools                69.0.3             pyhd8ed1ab_0    conda-forge
tk                        8.6.13          noxft_h4845f30_101    conda-forge
tzdata                    2023d                h0c530f3_0    conda-forge
wheel                     0.42.0             pyhd8ed1ab_0    conda-forge
xz                        5.2.6                h166bdaf_0    conda-forge
```

kbseah commented:
@BiocondaBot please fetch artifacts

kbseah commented:
@msantorsola one issue I found is that the epigen scripts (including simulate_data.py) do not have a "shebang" line at the head of the file. Therefore, even though they have been set to executable mode in the build script with `chmod +x`, the shell does not know that it should use the python interpreter to run them.

You'll need to patch the python script files to include the following (or similar) as the first line (no empty lines preceding) in each of the scripts:
```bash
#!/usr/bin/env python3
```

msantorsola commented:
> Package(s) built on Azure are ready for inspection



> @msantorsola one issue I found is that the epigen scripts (including simulate_data.py) do not have a "shebang" line at the head of the file. Therefore, even though they have been set to executable mode in the build script with `chmod +x`, the shell does not know that it should use the python interpreter to run them.
> 
> You'll need to patch the python script files to include the following (or similar) as the first line (no empty lines preceding) in each of the scripts:
> 
> ```shell
> #!/usr/bin/env python3
> ```

Ok, I will fix this in a fork repo

kbseah commented:
Actually you can write a patch file to fix it during the build process, without having to fork the project, see https://bioconda.github.io/contributor/guidelines.html#patching

msantorsola commented:
> Actually you can write a patch file to fix it during the build process, without having to fork the project, see https://bioconda.github.io/contributor/guidelines.html#patching

Dear @kbseah, I added the patch file and fixed the test errors. Anyway, when locally testing with:

`generate_genotype_corpus.py --corpus-id 21 --pop CEU --inds 1000 --chroms 21 --compress`

I get this error:

```
Traceback (most recent call last):
  File "/home/santorsola/.conda/envs/test_epigen/bin/generate_genotype_corpus.py", line 115, in <module>
    from utils.genotype_corpus_generator import GenotypeCorpusGenerator as GenCorGen
  File "/home/santorsola/.conda/envs/test_epigen/share/epigen/utils/genotype_corpus_generator.py", line 32, in <module>
    import matplotlib.pyplot as plt
ModuleNotFoundError: No module named 'matplotlib'
```

The meta.yaml seems not to work as it is

```
requirements:
  run:
    - python >=3.3
    - matplotlib-base
    - scipy >=1.3.1
    - numpy >=1.17.3
  host:
    - python >=3.3
    - pip
```

kbseah commented:
I'm afraid I wasn't able to reproduce the error. I installed the built package above and matpotlib-base was in the Conda environment. There was a different issue, that the paths to the `corpora` folder have to be updated.

msantorsola commented:
> I'm afraid I wasn't able to reproduce the error. I installed the built package above and matpotlib-base was in the Conda environment. There was a different issue, that the paths to the `corpora` folder have to be updated.

Dear @kbseah I modified the `build.sh` to add the path to corpora, anyway I still get the following error when testing:

```
$ simulate_data.py --sim-ids 0 --corpus-id 21 --pop ASW --inds 10 --snps 10 --disease-snps 156 3 1076 --model models/param_model.xml
Traceback (most recent call last):
  File "/home/santorsola/.conda/envs/test_epigen/bin/simulate_data.py", line 250, in <module>
    from utils.data_simulator import DataSimulator as DataSim
  File "/home/santorsola/.conda/envs/test_epigen/share/epigen/utils/data_simulator.py", line 31, in <module>
    from scipy import stats
ModuleNotFoundError: No module named 'scipy'
```

after building the recipe as follows:

`bioconda-utils build recipes config.yml --git-range master --package epigen`

How can I fix this issue?

kbseah commented:
hi @msantorsola that's strange, perhaps it has to do with the installation? i was able to run `simulate_data.py -h` successfully with the following steps:

First I git fetched your fork of `bioconda-recipes`, cd'd to the `bioconda-recipes` folder and checked out the `epigen` branch

```bash
conda activate ~/mambaforge/envs/bioconda # environment where i have installed `bioconda-utils
bioconda-utils build --packages epigen # recipes and config.yml implicit, only build epigen package
mamba create -p ./test-epigen
conda activate ./test_epigen
mamba install -c ~/mambaforge/envs/bioconda/conda-bld epigen # the built packages should be located in the `conda-bld` subfolder of the environment folder
simulate_data.py -h # no error messages
```

I am using bioconda-utils v2.11.1

msantorsola commented:
> hi @msantorsola that's strange, perhaps it has to do with the installation? i was able to run `simulate_data.py -h` successfully with the following steps:
> 
> First I git fetched your fork of `bioconda-recipes`, cd'd to the `bioconda-recipes` folder and checked out the `epigen` branch
> 
> ```shell
> conda activate ~/mambaforge/envs/bioconda # environment where i have installed `bioconda-utils
> bioconda-utils build --packages epigen # recipes and config.yml implicit, only build epigen package
> mamba create -p ./test-epigen
> conda activate ./test_epigen
> mamba install -c ~/mambaforge/envs/bioconda/conda-bld epigen # the built packages should be located in the `conda-bld` subfolder of the environment folder
> simulate_data.py -h # no error messages
> ```
> 
> I am using bioconda-utils v2.11.1

Dear @kbseah bioconda-utils v2.11.1 and mamba fixed the issue and the local installation is successful also for me!
